### PR TITLE
Lots of Invoice work:

### LINKED ISSUE #94: Let patient pay bill in full - Patient
A patient should be able to pay their bill, however only pay the full amount.

### LINKED ISSUE #105: D/N create invoice following an appointment - D/N
Invoice is already created when booking appointment. This will be updating that invoice that exists where needed with, starttime, endtime,  is_paid, description, and charge.

### PR DESCRIPTION
**Pull request by Austish101**
- Dr/Nurse can update invoices - closes #105 
- Client viewer and payment is back - closes #94 
- Combined servlets InvoiceViewer, ClientInvoiceViewer, EventViewer

### CODE REVIEW

File: esd-group-3/src/java/com/InvoiceViewerServlet.java [hunk 1]
```diff
@@ -23,49 +26,155 @@
...
+        int role = (Integer) loginSession.getAttribute("role");
+        
+        // dr/nurse - to ViewAppointments.jsp
+        if (role == 1 || role == 2) {
+            request.getRequestDispatcher("pages/ViewAppointments.jsp").include(request, response); 
+            request.setAttribute("todaydate", LocalDate.now().toString());
+            request.setAttribute("maxdate", LocalDate.now().plusYears(1).toString());
+            request.setAttribute("onemonth", LocalDate.now().plusMonths(1).toString());
+            request.setAttribute("minusyear", LocalDate.now().minusYears(1).toString());
+            // request.getRequestDispatcher("pages/ViewAppointments.jsp").forward(request, response);
```
Comment (conranpearce) [inline esd-group-3/src/java/com/InvoiceViewerServlet.java]: Small but remove comment

File: esd-group-3/src/java/com/InvoiceViewerServlet.java [hunk 2]
```diff
@@ -23,49 +26,155 @@
...
+            // request.getRequestDispatcher("pages/ViewAppointments.jsp").forward(request, response); 
+        }
+        // client - to PayInvoices.jsp
+        else if (role == 3) {
+            request.getRequestDispatcher("pages/PayInvoices.jsp").include(request, response);
+            request.setAttribute("todaydate", LocalDate.now().toString());
+            request.setAttribute("maxdate", LocalDate.now().plusYears(1).toString());
+            request.setAttribute("onemonth", LocalDate.now().plusMonths(1).toString());
+            request.setAttribute("minusyear", LocalDate.now().minusYears(1).toString());
+            // request.getRequestDispatcher("pages/PayInvoices.jsp").forward(request, response);
```
Comment (conranpearce) [inline esd-group-3/src/java/com/InvoiceViewerServlet.java]: remove

File: esd-group-3/src/java/com/InvoiceViewerServlet.java [hunk 3]
```diff
@@ -23,49 +26,155 @@
...
+        }
+        // admin - to ViewTurnover.jsp
+        else if (role == 4) {
+            request.getRequestDispatcher("pages/ViewTurnover.jsp").include(request, response);  
+            request.setAttribute("todaydate", LocalDate.now().toString());
+            request.setAttribute("maxdate", LocalDate.now().plusYears(1).toString());
+            request.setAttribute("onemonth", LocalDate.now().plusMonths(1).toString());
+            request.setAttribute("minusyear", LocalDate.now().minusYears(1).toString());
+            // request.getRequestDispatcher("pages/ViewTurnover.jsp").forward(request, response);
```
Comment (conranpearce) [inline esd-group-3/src/java/com/InvoiceViewerServlet.java]: remove

File: esd-group-3/src/java/com/InvoiceViewerServlet.java [hunk 4]
```diff
@@ -23,49 +26,155 @@
...
-            request.setAttribute("turnover", turnover);
-            request.getRequestDispatcher("invoiceViewer.jsp").forward(request,response);
+            
+//            String[][] nameArray = new String[operationsArray.size()][2];
+//            int j = 0;
+//            for(Operation i:operationsArray){
+//                nameArray[j][1] = (String) i.getClientFullNameFromId(dbcon);
+//                nameArray[j][2] = capitalizeWord(i.getRoleFromId(dbcon))  + " " + i.getEmpLastNameFromId(dbcon);
+//                j++;
+//            }
```
Comment (conranpearce) [inline esd-group-3/src/java/com/InvoiceViewerServlet.java]: remove comments

File: esd-group-3/src/java/com/InvoiceViewerServlet.java [hunk 5]
```diff
@@ -23,49 +26,155 @@
...
+            
+            // send data to correct page
+            // dr/nurse - to ViewAppointments.jsp
+            if (role == 1 || role == 2) {
+                request.setAttribute("message", "Data Loaded Successfully");
+                request.setAttribute("data", operationsArray); // Will be available as ${data}
+                // request.setAttribute("names", nameArray);
```
Comment (conranpearce) [inline esd-group-3/src/java/com/InvoiceViewerServlet.java]: remove

File: esd-group-3/src/java/com/InvoiceViewerServlet.java [hunk 6]
```diff
@@ -23,49 +26,155 @@
...
+                request.getRequestDispatcher("pages/PayInvoices.jsp").forward(request,response);
+                response.sendRedirect("pages/PayInvoices.jsp");
+            }
+            else if (role == 4) {
+                request.setAttribute("message", "Error - SQL Exception"); // Will be available as ${message}
+                request.getRequestDispatcher("pages/ViewTurnover.jsp").forward(request,response);
+                response.sendRedirect("pages/ViewTurnover.jsp");
+            }
+            
+            // response.sendRedirect("invoiceViewer.jsp");
```
Comment (conranpearce) [inline esd-group-3/src/java/com/InvoiceViewerServlet.java]: remove

File: esd-group-3/web/WEB-INF/web.xml [hunk 7]
```diff
@@ -115,9 +115,8 @@
         <servlet-class>com.ViewReferralsServlet</servlet-class>
     </servlet>
     <servlet>
-        <display-name>DisplayEventsServlet</display-name>
-        <servlet-name>DisplayEventsServlet</servlet-name>
-        <servlet-class>com.DisplayEventsServlet</servlet-class>
+        <servlet-name>UpdateInvoiceServlet</servlet-name>
+        <servlet-class>com.UpdateInvoiceServlet</servlet-class>
```
Comment (conranpearce) [inline esd-group-3/web/WEB-INF/web.xml:119]: DisplayEventsServlet removed from web.xml but the file still exists and is called in admin_home and client_home?

**Review by conranpearce — CHANGES_REQUESTED** [review]
Line 46 in ClientInvoiceViewerServlet"
operationsArray = operationsCaller.retrieveAllClientOperations(dbcon, userid, unpaid);
Is giving me an error for "retrieveAllClientOperations"?

**Comment by Austish101** [comment]
I love it when merge doesn't work, shouldn't be too bad to clean up.

**Review by ItsmeHJB — APPROVED** [review]
No issues found whilst testing

**Review by sam-kirk — CHANGES_REQUESTED** [review]
<div class="center">
            <div class="card patients-card">

^ add to viewAppointment.jsp after about line 109

<hr>

^ add to client_home.jsp

**Review by sam-kirk — APPROVED** [review]

**Review by conranpearce — APPROVED** [review]

**Pull request merged**